### PR TITLE
Time - Alicia

### PR DESCRIPTION
**Pull request by aecombs**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | I struggled with understanding what justified being a class or having inheritance. I also wanted to over-design the project in the beginning and that slowed me down, as well as overwhelmed me.
What was a design decision you made that changed over time over the project? | I ended up changing how I used Date objects. I originally figured that I could just receive all dates as arrays that I would then change into Date objects, but that just seemed weird. Also, I created a Rooms class after much back and forth over it.
What was a concept you gained clarity on, or a learning that you'd like to share? | I think my biggest take away is that it's okay for me to take it one step at a time. Looking ahead didn't help me much. It meant that I was completing wave 1 and wave 2 at the same time, and the complexity was absolutely overwhelming.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | I didn't write too many nominal tests, but I did when wanting to calculate the cost of the hotel block rooms. It would be considered nominal because I had hardcoded a magic number after the "must_equal" bit.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | I wrote edge cases for situations in which an invalid argument was being passed. For instance, creating a DateRange and not receiving Date objects, or creating a HotelBlock object and receiving a non-array, or an array of non-Rooms. I'd say they were edge cases because it's not how a user is expected to use the program.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | Writing in pseudocode kind of changed my life! It made it so that I could break down problems easily and think about specifics on how I'd implement my designs after getting it out of my head. Writing the tests before writing my code was much more challenging, however; I typically would get excited about how I could solve my problems that I'd want to code it out right away.

### CODE REVIEW

**Comment by aecombs** [comment]
Note to whomever grades this:

A note on my Simplecov test coverage: no matter how many tests I added or how thorough I tried to be with each test, I couldn't get simplecov to acknowledge anything within my methods. Because of this, it only shows 30% coverage on most of my files -- which encompasses all the the method signatures and what's in the initialize method. Nothing inside any of the methods. I wanted to explicitly mention this so that I don't get dinged on this bug.

File: test/hotel_block_test.rb [hunk 1]
```diff
@@ -0,0 +1,108 @@
+require 'test_helper'
```
Comment (jmaddox19) [inline test/hotel_block_test.rb:1]: I needed to change this to match your other tests before I could run rake. 
```suggestion
require_relative 'test_helper'
```

File: lib/date_range.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      return self.create_days_array.include?(date)
+    end
+
+    def overlap?(other_date_range)
+      days = self.create_days_array
+      other_days = other_date_range.create_days_array
+    
+      return false if (days.first === other_days.last || days.last === other_days.first) || days.difference(other_days).length === days.length
+      return true
+    end
```
Comment (jmaddox19) [inline lib/date_range.rb:39]: I love all this! It's really easy to read and great that the logic is separated out into smaller methods.

**Comment by jmaddox19** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | ✔️ Though I'm concerned that you might not have been running tests as your worked because tests didn't run for me without a fix
Practices git with at least 15 small commits and meaningful commit messages | ✔️

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | ✔️
All of the given tests run and pass | Not until I made the change noted in my inline comment
A test coverage tool is installed and used, and shows 95% test coverage | I'm not seeing coverage output. Not sure why :( 

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | ✔️
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | ✔️
calculates the total price for a reservation | ✔️
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | ✔️
creates a block of rooms | ✔️
reserves a room from a block | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections | ✔️
| Yellow (Approaches Standards) | 9-13 total in all sections |
| Red (Not at Standard) | 0-8 total in all sections, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |

### Additional Feedback



Great work overall! You've built your first project with minimal starting code. This represents an incredible milestone in your journey, and you should be proud of yourself!

I am particularly impressed by the way that you separated your code into small methods and the way you tested everything so thoroughly!


I do see some room for improvement around ensuring that your tests all run when submitting the project. Running "rake" on the command line was not working for me. (See inline comments for more context on why.) It especially concerns me that the tests weren't working because it causes me to wonder if you were running the tests frequently while you were writing code. I'm hopeful you were running your tests and this was somehow working for you even though it wasn't for me.


I also want to respond to your note about writing tests first slowing you down. I would say in general it is normal for writing thorough tests to take you as long (if not longer) than writing the source code. However, I want to name that writing tests before writing source code is something worth trying a couple times but you are welcome to decide it isn't serving you well and write tests after writing code instead. :) 

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | 
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
